src/runtime/pipeline/pipeline_struct.h

```diff
@@ -134,37 +220,316 @@ struct OutputMap {
+      LOG(FATAL) << "The Configuration loading not finish yet.";
+    }
+    module_input_output_binding_total_num = 0;
     for (auto mod_output : config) {
-      num_output += mod_output.second.GetGlobalOutputNum();
+      // Get the numbers of binding of input and output.
+      module_input_output_binding_total_num += mod_output.second.GetInputOutputBindingNum();
+      // Use global output index as key to create a mapping of global index and module output.
+      const std::vector<GlobalOutputPair>& global_output =
+          mod_output.second.GetGlobalConfigOutputBindings();
+
+      for (auto output : global_output) {
+        global_output_map[output.global_output_idx] =
+            ModuleOutputPair(mod_output.first, output.mod_output_idx);
+      }
     }
-    return num_output;
+    return;
+  }
+  /*!
+   * \brief Create a pipeline config from JSONReader.
+   * \param reader Json reader.
+   */
+  void Load(dmlc::JSONReader* reader) {
+    reader->BeginArray();
+    while (reader->NextArrayItem()) {
+      std::string key;
+      reader->BeginObject();
+      int mod_idx = -1;
+      ConfigOutputBindings output;
+      std::string dev;
+      while (reader->NextObjectItem(&key)) {
+        if (key == "mod_idx") {
+          reader->Read(&mod_idx);
+        } else if (key == "dev") {
+          reader->Read(&dev);
+        } else if (key == "output") {
+          reader->Read(&output);
+        } else {
+          LOG(FATAL) << "do not support key " << key;
+        }
+      }
+      ICHECK(mod_idx >= 0) << "Invalid mod_idx value " << mod_idx;
+      // Check if the output is successfully read.
+      ICHECK(!output.Empty()) << "Invalid output binding result.";
+      Insert(mod_idx, output);
+    }
+    // Call this function after "config" loading finished.
+    ParseConfiguration(config);
+  }
+};
+/*
+ *\brief Runtime of backend.
+ */
+class BackendRuntime {
+ private:
+  /*\brief The index of runtime indicate the position in the pipeline.*/
+  int runtime_idx;
+  /*\brief The Runtime module of a backedn graph executor.*/
+  Module module;
+  /*!
+   *\brief To transfer data between two different backends, we need a local
+   * tensor variable as a medium. This variable is a mapping of input data and local
+   * data.
+   */
+  std::unordered_map<DLTensor*, DLTensor*> input_tensor_local_copy;
+  /*!\brief The packed functions.*/
+  tvm::runtime::PackedFunc run;
+  tvm::runtime::PackedFunc set_input;
+  tvm::runtime::PackedFunc get_input;
+  tvm::runtime::PackedFunc get_output;
+  tvm::runtime::PackedFunc get_num_output;
+  tvm::runtime::PackedFunc get_num_inputs;
+  tvm::runtime::PackedFunc get_input_index;
+  /*!\brief The new DLTensor have same shape, data type with a existing DLTensor.*/
+  DLTensor* CreateFromDLTensor(const DLTensor* from) {
+    DLTensor* ret = NULL;
+    TVMArrayAlloc(from->shape, from->ndim, from->dtype.code, from->dtype.bits, from->dtype.lanes,
+                  kDLCPU, 0, &ret);
```

Comment:
Why CPU?

Reply:
this is to help cross device data copy, when doing two non-cpu device NDArray copy , in current NDArray::CopyFromTo logic,  the "from" device DeviceAPI will get use to for the data copy, but the memory reference in "to" device NDArray may only available by using "to" device DeviceAPI, that means to finish a cross device copy we need to involve both "from" and "to" device DeviceAPI,  there solution here is to use a "cpu" NDArray as a medium to first copy data to CPU NDArray then copy the cpu NDArray to target device.

Reply:
I see. Then better rename or inline this function, as it is used only once. If you prefer to rename, something like `CopyDLTensorToCPU` might be better.

Reply:
fixed.

Reply:
fixed.